pkg/pillar/cmd/zedrouter/metadata_handlers.go

```diff
@@ -0,0 +1,713 @@
+
+		patchID := chi.URLParam(r, "patch")
+		if patchID == "" {
+			sendError(w, http.StatusNoContent, "patch in route is missing")
+			return
+		}
+		e := types.FindPatchEnvelopeById(envelopes, patchID)
+		if e != nil {
+			path, err := os.MkdirTemp("", "patchEnvelopeZip")
+			if err != nil {
+				sendError(w, http.StatusInternalServerError,
+					fmt.Sprintf("failed to create temp dir %v", err))
+				return
+			}
+			zipFilename, err := types.GetZipArchive(path, *e)
+
+			if err != nil {
+				sendError(w, http.StatusInternalServerError,
+					fmt.Sprintf("failed to archive binary blobs %v", err))
+				return
+			}
+
+			http.ServeFile(w, r, zipFilename)
+
+			err = os.Remove(zipFilename)
+			if err != nil {
+				sendError(w, http.StatusInternalServerError,
+					fmt.Sprintf("failed to delete archive %v", err))
+				return
+			}
+			return
+		}
+		http.Error(w, http.StatusText(http.StatusUnauthorized), http.StatusUnauthorized)
+		return
+	}
+}
+
+// HandlePatchFileDownload serves binary artifact of specified patch envelope to app
+// instance. Patch envelope id and file name is specified in URL.
+func HandlePatchFileDownload(z *zedrouter) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		// WithPatchEnvelopesByIP middleware returns envelopes which are more than 0
+		envelopes := r.Context().Value(patchEnvelopesContextKeyType).([]types.PatchEnvelopeInfo)
+
+		patchID := chi.URLParam(r, "patch")
+		if patchID == "" {
+			sendError(w, http.StatusNotFound, "patch in route is missing")
+			return
+		}
+		fileName := chi.URLParam(r, "file")
+		if fileName == "" {
+			sendError(w, http.StatusNotFound, "file in route is missing")
+			return
+		}
+
+		e := types.FindPatchEnvelopeById(envelopes, patchID)
+		if e != nil {
+			if idx := types.CompletedBinaryBlobIdxByName(e.BinaryBlobs, fileName); idx != -1 {
+				http.ServeFile(w, r, e.BinaryBlobs[idx].Url)
+				return
+			} else {
+				sendError(w, http.StatusNotFound, "file is not found")
+				return
+			}
+		}
+
+		sendError(w, http.StatusNotFound, "patch is not found")
+	}
+}
+
+// WithPatchEnvelopesByIP is a middleware for Patch Envelopes which adds
+// to a context patchEnvelope variable containing available patch envelopes
+// for given IP address (it gets resolved to app instance UUID)
+// in case there is no patch envelopes available it returns StatusNoContent
+func WithPatchEnvelopesByIP(z *zedrouter) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			peObj, err := z.subPatchEnvelopeInfo.Get("zedagent")
+			if err != nil {
+				sendError(w, http.StatusNoContent, "Cannot get patch envelope subscription")
+				return
+			}
+
+			pe := peObj.(types.PatchEnvelopes)
+
+			remoteIP := net.ParseIP(strings.Split(r.RemoteAddr, ":")[0])
+			anStatus := z.lookupAppNetworkStatusByAppIP(remoteIP)
```

Comment:
Do you check if the given app instance (caller of API) is part of the allowed list before reaching here?

Reply:
I follow same logic as other handlers in resolving remoteAddr to uuid, then check if there are any patch envelopes available for given app instance uuid and return them